src/Futhark/CodeGen/Backends/GenericWASM.hs

```diff
@@ -0,0 +1,400 @@
+    free() { this.ffree(this.ctx, this.ptr); }
+    shape() {
+      var s = this.fshape(this.ctx, this.ptr) >> 3;
+      return Array.from(HEAP64.subarray(s, s + this.dim));
+    }
+    toTypedArray(dims = this.shape()) {
+      console.assert(dims.length === this.dim, "dim=%s,dims=%s", this.dim, dims.toString());
+      var length = Number(dims.reduce((a, b) => a * b));
+      var v = this.fvalues(this.ctx, this.ptr) / this.heap.BYTES_PER_ELEMENT;
+      return this.heap.subarray(v, v + length);
+    }
+    toArray() {
+      var dims = this.shape();
+      var ta = this.toTypedArray(dims);
+      return (function nest(offs, ds) {
+        var d0 = Number(ds[0]);
+        if (ds.length === 1) {
+          return Array.from(ta.subarray(offs, offs + d0));
+        } else {
+          var d1 = Number(ds[1]);
+          return Array.from(Array(d0), (x,i) => nest(offs + i * d1, ds.slice(1)));
+        }
+      })(0, dims);
+    }
+  }
+  Module['FutharkArray'] = FutharkArray;
+  |]
+
+classFutharkContext :: [JSEntryPoint] -> String
+classFutharkContext entryPoints =
+  unlines
+    [ classDef,
+      constructor entryPoints,
+      getFreeFun,
+      getEntryPointsFun,
+      getErrorFun,
+      unlines $ map toFutharkArray arrays,
+      unlines $ map jsWrapEntryPoint entryPoints,
+      endClassDef,
+      "Module['FutharkContext'] = FutharkContext;"
+    ]
+  where
+    arrays = filter isArray typs
+    typs = nub $ concatMap (\jse -> parameters jse ++ ret jse) entryPoints
+
+classDef :: String
+classDef = "class FutharkContext {"
+
+endClassDef :: String
+endClassDef = "}"
+
+constructor :: [JSEntryPoint] -> String
+constructor jses =
+  T.unpack
+    [text|
+  constructor(num_threads) {
+    this.cfg = _futhark_context_config_new();
+    if (num_threads) _futhark_context_config_set_num_threads(this.cfg, num_threads);
+    this.ctx = _futhark_context_new(this.cfg);
+    this.entry_points = {
+      ${entries}
+    };
+  }
+  |]
+  where
+    entries = T.pack $ intercalate "," $ map dicEntry jses
+
+getFreeFun :: String
+getFreeFun =
+  T.unpack
+    [text|
+  free() {
+    _futhark_context_free(this.ctx);
+    _futhark_context_config_free(this.cfg);
+  }
+  |]
+
+getEntryPointsFun :: String
+getEntryPointsFun =
+  T.unpack
+    [text|
+  get_entry_points() {
+    return this.entry_points;
+  }
+  |]
+
+getErrorFun :: String
+getErrorFun =
+  T.unpack
+    [text|
+  get_error() {
+    var ptr = _futhark_context_get_error(this.ctx);
+    var len = HEAP8.subarray(ptr).indexOf(0);
+    var str = String.fromCharCode(...HEAP8.subarray(ptr, ptr + len));
+    _free(ptr);
+    return str;
+  }
+  |]
+
+dicEntry :: JSEntryPoint -> String
+dicEntry jse =
+  T.unpack
+    [text|
+  '${ename}' : [${params}, ${rets}]
+  |]
+  where
+    ename = T.pack $ name jse
+    params = T.pack $ show $ parameters jse
+    rets = T.pack $ show $ ret jse
+
+jsWrapEntryPoint :: JSEntryPoint -> String
+jsWrapEntryPoint jse =
+  T.unpack
```

Comment:
Why are we unpacking these instead of keeping them as `Text`s?

Reply:
My initial implementation worked off of strings. There are a few points when I have to combine strings. In some places they are 'Text' and others they are 'String'. So I just decided to convert everything to 'String' so that I didn't have any weird inconsistencies later.

Reply:
Fine, but if it's not too much bother, then Text is (much) faster.  Probably not the bottleneck in such a slow compiler as Futhark, but I've been meaning to do a String exorcism in general.  That can be done later, I suppose.